src/Storages/StorageMerge.cpp

```diff
@@ -496,17 +546,28 @@ DatabaseTablesIteratorPtr StorageMerge::getDatabaseIterator(ContextPtr local_con
         throw;
     }
 
-    auto database = DatabaseCatalog::instance().getDatabase(source_database);
+    DatabaseTablesIterators database_table_iterators;
 
-    auto table_name_match = [this](const String & table_name_) -> bool
+    auto databases = DatabaseCatalog::instance().getDatabases();
+
+    for (const auto & db : databases)
```

Comment:
It seems it might be really reducing performance in case no regexp is used for database_name because `getDatabaseIterators` is used all the time and it will always traverse all existing databases more than once for each read :( 

May be it is better to make a way in table startup to define whether passed database_name is a regexp or not. At least most common case - if it contains only alphanumeric + underscore `_` + `-`. Or you have some ideas?

Also there is a strange case when somebody can have a database name which can be considered as regexp but actually it is not. Like dots or `+` in database name. Though it is really unlikely so it does not matter probably.

Reply:
Or maybe we can make it as an `optional` member variable of StorageMerge?

Reply:
Or maybe, we can add an optional arguments to indicate whether the database_name is a regexp or not, the default is not?

Reply:
I think you are right, since currently we use full match to match database name and table name, so if it contains only alphanumeric, `_` and `-`, it can not be a regexp?

Reply:
If there are, for example, dots in database name, it might be either regexp or not, so it is not a good solution.
There has been proposed an idea to make some identifier in table definition `CREATE TABLE merge_table (...) ENGINE = Merge(REGEXP(database_name), 'table_regexp')`, because it was discussed that it is not ok to have different match algorithms for database_name and for table_name (FullMatch vs Match). 
They should both be the same - just Match, but making Match instead of FullMatch will break previously created Merge table as without FullMatch it will start to match more tables. So the solution is to introduce a `REGEXP(database_name)` expression, and Merge table will be created either with it (then database name is treated as regexp), or without it (then it is not regexp).

Reply:
>  CREATE TABLE merge_table (...) ENGINE = Merge(REGEXP(database_name), 'table_regexp')

This look better, I will try to implement it.

BTW, we still should use Match for both database_name and table_name?

Reply:
Yes, it should me Match for both database_name and table_name, now as I see it is FullMatch for database name and Match for table_name.

Reply:
But database_name can also be `REGEXP(...)`, while it's unusual.

Reply:
I think it 100% unlikely that this will happen, it would be really strange to have such database_name.

Reply:
My idea to implement `REGEXP(database_name` is treat it as `ASTFunction` and exact it's argument. Does it is OK?

Reply:
I think Ok.